ARCHITECTURE.md

```diff
@@ -0,0 +1,94 @@
+
+More specifically, kedro-viz consumes pipeline data in a specific object format that includes all pipelines, edges an nodes of the Kedro project, meanwhile the Kedro-viz API layer makes request to the Kedro project for further details of a node object, which will be requested and displayed on user clicking a node on the flowchart. (One important thing to note is the concept of a 'node' in Kedro-viz, which is different from the concept of 'node' in Kedro. A 'node' on Kedro-viz referes to an element for display on the flowchart, which could be either a 'database', 'parameter' or 'function' element.)
+
+Kedro-viz is made up of two main parts: the front-end web-app UI that is built on React, and an API layer that pulls data from the Kedro project on which Kedro-viz is running on. 
+
+Currently, the API has 2 endpoints: `/pipeline/<id>` and `/node/<id>`. Each endpoint is called via actions within the app. 
+
+Kedro-viz can exist either as a standalone web-app (via spinning up the web-app from a bash command with data consumed from a running Kedro project), or as a react component that can be imported in any external web-app (consuming data from a JSON file instead of a running Kedro project).  
+
+# Our tools
+This section outlines the third party tools that we heavily utilize within Kedro-viz. 
+
+### Redux
+We use redux to manage our data flow and state manegement within our web-app, utilizing the redux-toolkit (https://redux-toolkit.js.org/) for all redux-related configuration. 
+
+Upon firing an action to obtain the pipeline data (either from the `/pipeline` endpoint or via a JSON file), the pipeline data is further prepared and stored within the redux store as a single source of truth within the app. 
+
+All data-fetching from the API and updates to the state within the app are strictly done via actions, while all data are consumed by components either directly through the state or selectors. 
+
+### Web workers
+Kedro-viz utilizes web workers to asynchronously perform time-consuming calculations (e.g. for instance the dagre/newgraph layout calculation for the flowchart) in a separate CPU thread, in order to prevent it from blocking other operations on the main thread (e.g. CSS transitions and other state updates).
+
+The layout calculation as performed by the web-worker is masked by the spinning loader to improve the user experience in larger graphs.
+
+### SCSS imports
+Kedro-Viz utilizes SCSS imports to utilize and access mixins and variables in our stylesheets. 
+
+### Jest & React testing library
+All tests within Kedro-viz are set up with Jest, with some tests set up using React testing library to utilize their `fireEvent` function. 
+
+# Codemap / features
+This section outlines the important directories, its related purpose and data requirements within the app. 
+
+## data normalisation & Redux store
```

Comment:
We want to keep this as brief and high-level as possible - we should avoid describing how each file works - you can use in-file comments for this!

Instead, we want a brief, high-level overview of the data flow, specifically focused on why we do things the way we do. e.g. something like:

> On initialisation, the app uses a string token (e.g. 'json' or 'animals' to determine the data source - you can find the details about which identifier refers to which data source in [CONTRIBUTING.md](https://github.com/quantumblacklabs/kedro-viz/blob/main/CONTRIBUTING.md#data-sources). The application then [manually normalises this data](https://github.com/quantumblacklabs/kedro-viz/blob/main/src/store/normalize-data.js), in order to [make immutable state updates as performant as possible](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape).
> Next, the app [initialises the Redux data store](https://github.com/quantumblacklabs/kedro-viz/blob/main/src/store/initial-state.js), by merging together initial defaults, `window.localStorage` data, and the normalized pipeline data. The initial state data is separated into two: Pipeline and non-pipeline state. This is because the non-pipeline state should persist if the pipeline data is reset - e.g. if the user selects a new top-level pipeline.

or

> When loading asynchronous data (specified using the `json` identifier), the app first loads the `/api/main` endpoint. This is because it contains the list of available top-level pipelines, which is necessary to check before loading others. If another pipeline's data is required (which happens under certain conditions, e.g. [...]), it then loads that pipeline's data from `/api/pipeline/[ID]` and replaces the app state with this new data. Conversely, when the user selects a node on the graph, if this node has not previously been selected then the app will request data for this node from `/api/nodes/[ID]`.

You could write this a lot more accurately and informatively, this is just a couple of examples of the top of my head to give an indication of the level of depth that we should aim for. Focus on _what_ and _why_, not _how_. If you are unsure _why_ we do something the way we do, please ask! :)

Reply:
I've had another look at [an example architecture doc](https://github.com/rust-analyzer/rust-analyzer/blob/d7c99931d05e3723d878bea5dc26766791fa4e69/docs/dev/architecture.md) and it does have specific description of parts of the codebase, but more describing sections/directories rather than individual files. Maybe we should approach it like that. We should decide what the best taxonomy/breakdown would be, and how granular we need to get. Ideally as little granularity as possible, while still being descriptive.

Reply:
Thanks @richardwestenra , that's a great doc ( which the same doc has been heavily referenced for the structure and format of this doc too :D )

I have referenced your suggested paragraph in the newly added dataflow section (which also includes the data-flow diagram that we flushed out). Will let you know when I have pushed it and let me know your thoughts!

Reply:
I'm added a new data flow section that incorporates the ideas you suggested above.